modules/afprog/afprog.c

```diff
@@ -38,6 +38,33 @@
 #include <unistd.h>
 #include <signal.h>
 
+typedef struct _AFProgramReloadInfo
+{
+  LogWriter * writer;
+  pid_t       pid;
+} AFProgramReloadInfo;
+
+static inline void
+terminate_descendants_by_pid(const pid_t pid)
+{
+  msg_verbose("Sending descendant program group a TERM signal",
+              evt_tag_int("child_pid", pid),
+              NULL);
+
+  pid_t pgid = getpgid(pid);
+  if (pgid != -1)
+    killpg(pgid, SIGTERM);
+}
+
+static inline void
+afprogram_reload_info_free(AFProgramReloadInfo * const reload_info)
+{
+  child_manager_unregister(reload_info->pid);
+  terminate_descendants_by_pid(reload_info->pid);
+  log_pipe_unref(reload_info->writer);
+  g_free(reload_info);
+}
+
```

Comment:
Fixed in 459ffeec83e97ce636a2b0ba284d4ff6f23c68b1.

The new functions are `_deinit` and `_free`, because IMHO this fits to the syslog-ng terminology.